File: 백준/Gold/1958. LCS 3/LCS 3.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var s1, s2, s3 string
	fmt.Fscanf(reader, "%s\n%s\n%s", &s1, &s2, &s3)
	var res = make([][][]float64, len(s1)+1)
	for i := 0; i <= len(s1); i++ {
		res[i] = make([][]float64, len(s2)+1)
		for j := 0; j <= len(s2); j++ {
			res[i][j] = make([]float64, len(s3)+1)
		}
	}
	for i := 1; i <= len(s1); i++ {
		for j := 1; j <= len(s2); j++ {
			for k := 1; k <= len(s3); k++ {
				if s1[i-1] == s2[j-1] && s2[j-1] == s3[k-1] && s1[i-1] == s3[k-1] {
					res[i][j][k] = res[i-1][j-1][k-1] + 1
				} else {
					res[i][j][k] = math.Max(res[i-1][j][k], math.Max(res[i][j-1][k], res[i][j][k-1]))
				}
			}
		}
	}
	fmt.Printf("%0.f\n", res[len(s1)][len(s2)][len(s3)])
}

File: 백준/Gold/1958. LCS 3/LCS 3.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
s1 = input().strip()
s2 = input().strip()
s3 = input().strip()
res = [[[0 for _ in range (len(s3) + 1)] for _ in range(len(s2) + 1)] for _ in range(len(s1) + 1)]
for i in range(1, len(s1)+1):
    for j in range(1, len(s2)+1):
        for k in range(1, len(s3)+1):
            if s1[i - 1] == s2[j - 1] == s3[k - 1]:
                res[i][j][k]=res[i-1][j-1][k-1]+1
            else:
                res[i][j][k]=max(res[i - 1][j][k],res[i][j - 1][k],res[i][j][k-1])
print(res[-1][-1][-1])